doc: Add Egress Gateway Getting Started Guide (#15661)
cilium-team/cilium: Documentation/gettingstarted/egress-gateway.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _egress_gateway:

**********************
Egress Gateway (beta)
**********************

Egress gateway provides a way to configure the egress policy. Without
this feature, pod traffic could only egress from the node it is running

> on and would be masqueraded with the node IP.

Reviewer: I would turn this other way around:
```
The egress gateway allows users to redirect egress pod traffic through
specific, gateway nodes. Packets are masqueraded to the gateway node IP.
```

Author: agreed, fixed.

This document explains how to configure Cilium to enable egress gateway
as well as how to configure egress NAT policy to route and SNAT the egress

> traffic for a specific workload.

Reviewer: ```suggestion
This document explains how to enable the egress gateway and configure
egress NAT policies to route and SNAT the egress traffic for a specific
workload.
```

Author: fixed.

.. include:: ../beta.rst

Configure Cilium
================

.. include:: k8s-install-download-release.rst

The Cilium Egress Gateway feature relies on :ref:`kubeproxy-free`,

> follow the guide to create a new deployment. The feature is disabled by default.

Reviewer: The kube-proxy-free guide instructs users to remove kube-proxy completely. I'm worried we might give the wrong impression here, that kube-proxy needs to be removed. Maybe we can just say it require our kube-proxy replacement (with a link) and add `kubeProxyReplacement=strict` in the command below.

> Enable the feature by setting the ``localRedirectPolicy`` value to ``true``.

Reviewer: ```suggestion
Enable the feature by setting the ``enableEgressGateway`` value to ``true``.
```

Author: fixed.

.. parsed-literal::

   helm install cilium |CHART_RELEASE| \\

>      --set enableEgressGateway=true

Reviewer: This Helm option is currently not in master. We need the docs PR for Cilium v1.10-rc1.
@anfernee Could you submit the Helm option in a separate PR? I think it's already in the e2e test PR, right?

Verify that Cilium agent pod is running.

.. code-block:: shell-session

    $ kubectl -n kube-system get pods -l k8s-app=cilium
    NAME           READY   STATUS    RESTARTS   AGE
    cilium-5ngzd   1/1     Running   0          3m19s

> Validate that the Cilium Egress NAT Policy CRD has been registered.

Reviewer: I would drop this or, better, move it to a troubleshooting section. For a getting started guide, I think we want to take the shortest route. If the CRD is missing it's a bug and therefore corner case.

Author: removed for now.

.. code-block:: shell-session

	   $ kubectl get crds
	   NAME                              CREATED AT
	   [...]

> 	   ciliumegressnatpolicy.cilium.io   2021-03-26T02:11:21Z

Reviewer 2: Additionally, please use consistent indentation levels throughout the document. This one in particular stands out with the rest, but most of your code blocks have 4 spaces when they should have 3, like for your `parsed-literal` block.

Reviewer 3: +1, sphinx is quite picky about the indentation unfortunately.

Author: fixed the consistency issue. thx!

Optionally, Egress Gateway support can be enabled by setting
``enable-egress-gateway`` to ``true`` in ConfigMap. By default, it is disabled.

To set ``enable-egress-gateway`` to ``true``, it requires
``enable-bpf-masquerade`` and ``kube-proxy-replacement``. It cannot be enabled

> with either of these 2 options set as ``false`` or ``disabled``.

Reviewer: I would drop this paragraph and just add a note on line 28 to say it also requires BPF masquerading.

Author: done.

Architecture
============

Overall, some nodes in the cluster will be assigned as gateway nodes. These gateway nodes not
only have a node IP, but also have additional pools of egress IPs attached to it. These egress IP
addresses can be associated with certain workloads (pods) that run in the cluster. When certain
packets are selected, instead of exiting the cluster immediately, they will be forwarded to a gateway
node and SNAT-ed with an egress IP there. A tunnel interface will be created when Egress Gateway is
enabled. The tunnel is for forwarding with Egress Gateway. All other traffic will keep their
own behavior based on which tunnel mode is selected.

The following 2 diagrams show the packet flow for both egress and return traffic when using direct
routing mode. Note that the return packet is not going over a tunnel, while forwarding traffic is. 

.. image:: images/cilium_egress_gateway_flow.png
   :scale: 70%

.. image:: images/cilium_egress_gateway_return_flow.png

>    :scale: 70%

Reviewer: This section is great! But it's usually something we include in the blog post for the new release (I'll ping you when we start writing that). If we include it in the documentation, then (1) it should probably be somewhere in Concepts instead and (2) that means we'll try to maintain it. I would just remove this and keep it on the side for the blog post.

Author: I think it makes sense. Removed for this guide.

Create an External Service
==========================

Create an nginx service on a Linux node that is external to the existing Kubernetes

> cluster, and use it as the destination of the egress traffic.

Reviewer: I understand this is useful to show that the received packets have been masqueraded to the expected egressIP, but most users testing this guide are likely to do so on managed Kubernetes platforms where reproducing this is not as straightforward.

Do we really need to show the source IP in the nginx logs? Could we instead show in some other way that the packets have gone through the egress gateway node as expected?

Author: I think this is just for an example of verification that is easy to understand and easy to use in practice.

Reviewer: I agree it's a good example, but how are users with e.g. a GKE cluster supposed to follow this guide?

Reviewer 4: They should be able to create a vm instance and install nginx. But you are right, this should be marked as optional. The user should be able to use their own external service to verify.

Reviewer: I'm fine if we mark that step as optional.

.. code-block:: shell-session

    $ # Install and start nginx
    $ sudo apt install nginx
    $ sudo systemctl start nginx

    $ # Make sure the service is started and listens on port :80
    $ sudo systemctl status nginx
    ● nginx.service - A high performance web server and a reverse proxy server
    Loaded: loaded (/lib/systemd/system/nginx.service; enabled; vendor preset: enabled)
    Active: active (running) since Sun 2021-04-04 21:58:57 UTC; 1min 3s ago
    [...]
    $ curl http://192.168.33.13:80  # Assume 192.168.33.13 is the external IP of the node
    [...]
    <title>Welcome to nginx!</title>
    [...]

Create Client Pods
==================

Deploy a client pod from which egress traffic is generated, and try to access previous
external service.

.. code-block:: shell-session

    $ kubectl run -it client --image=busybox
    If you don't see a command prompt, try pressing enter.

>     / # wget http://192.168.33.13:80 -o /dev/null

Reviewer: AFAIU, this will require users to switch between terminals to follow the guide. Could we instead use a proper Pod manifest? There are multiple examples in other guides (LRP guide).

EDIT: I see you're already using a manifest later on. Maybe we could use it sooner?

Verify access log from previous node that the request is coming from one of the node in
Kubernetes cluster.

.. code-block:: shell-session

    $ tail /var/log/nginx/access.log
    192.168.33.11 - - [04/Apr/2021:22:06:57 +0000] "GET / HTTP/1.1" 200 612 "-" "Wget/1.19.4 (linux-gnu)"

> In the previous example, the busybox is running on the node ``192.168.33.11``, so the result makes sense.

Reviewer: I think we can skip this first check. This is essentially testing Cilium works, which is out of scope for the guide IMO. If you really want to show before/after, I would configure everything, show before, add the label, show after.

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _egress_gateway:

**********************
Egress Gateway (beta)
**********************

Egress gateway provides a way to configure the egress policy. Without
this feature, pod traffic could only egress from the node it is running

> on and masqueraded with the node IP.

Reviewer 2: ```suggestion
on and would be masqueraded with the node IP.
```

This document explains how to configure Cilium to enable egress gateway
as well as how to configure egress NAT policy to route and SNAT the egress
traffic for a specific workload.

.. include:: ../beta.rst

Configure Cilium
================

.. include:: k8s-install-download-release.rst

The Cilium Egress Gateway feature relies on :ref:`kubeproxy-free`,

> follow the guide to create a new deployment. The beta feature is disabled by default.

Reviewer 2: ```suggestion
follow the guide to create a new deployment. The feature is disabled by default.
```
No need to remind that it is in beta mode here. Plus this is likely something we would forget to remove when moving out of beta.

Enable the feature by setting the ``localRedirectPolicy`` value to ``true``.

.. parsed-literal::

   helm install cilium |CHART_RELEASE| \\
     --set enableEgressGateway=true

Verify that Cilium agent pod is running.

> .. code-block:: bash

Reviewer 2: ```suggestion
.. code-block:: shell-session
```

    $ kubectl -n kube-system get pods -l k8s-app=cilium
    NAME           READY   STATUS    RESTARTS   AGE
    cilium-5ngzd   1/1     Running   0          3m19s

Validate that the Cilium Egress NAT Policy CRD has been registered.

> .. code-block:: bash

Reviewer 2: ```suggestion
.. code-block:: shell-session
```

	   $ kubectl get crds
	   NAME                              CREATED AT
	   [...]
	   ciliumegressnatpolicy.cilium.io   2021-03-26T02:11:21Z

Optionally, Egress Gateway support can be enabled by setting
``enable-egress-gateway`` to ``true`` in ConfigMap. By default, it is disabled.

To set ``enable-egress-gateway`` to ``true``, it requires

> ``enable-bpf-masqurade`` and ``kube-proxy-replacement``. It cannot be enabled

Reviewer 2: ```suggestion
``enable-bpf-masquerade`` and ``kube-proxy-replacement``. It cannot be enabled
```

> with these 2 options set as ``false`` or ``disabled``.

Reviewer 2: ```suggestion
with either of these 2 options set as ``false`` or ``disabled``.
```
What about KPR in partial mode?

Reviewer 2: What feature does it need exactly? Can kube-proxy replacement be used in partial mode, or does it have to be strict mode?

Author: This feature is designed and tested under kube-proxy-replacement strict mode. I don't know if we can support its functionality when kube-proxy is partially present. @anfernee could you comment on this?

Reviewer: `=partial` won't work with the egress gateway unless all KPR sub-features are enabled, so I think we should just recommend to set `=strict`.

Architecture
============

> Overall, some nodes in the cluster will be assigned as gateway nodes. These gateway nodes not
> only have node IP, but also have additional pools of egress IPs attached to it. These egress IP
> addresses can be associated with certain workloads (pods) that run in the cluster. When certain
> packets are selected, instead of exitting the cluster immediatly, they will be forwarded to gateway
> node and SNAT with egress IP there. A tunnel interface will be created when Egress Gateway is
> enabled. The tunnel is for the forwarding in Egress Gateway. All other traffic will keep their
> own behavior based on which tunnel mode is selected.

Reviewer 2: ```suggestion
Overall, some nodes in the cluster will be assigned as gateway nodes. These gateway nodes not
only have a node IP, but also have additional pools of egress IPs attached to it. These egress IP
addresses can be associated with certain workloads (pods) that run in the cluster. When certain
packets are selected, instead of exiting the cluster immediately, they will be forwarded to a gateway
node and SNAT-ed with an egress IP there. A tunnel interface will be created when Egress Gateway is
enabled. The tunnel is for forwarding with Egress Gateway. All other traffic will keep their
own behavior based on which tunnel mode is selected.
```

`some nodes in the cluster` as well as `When certain packets are selected` are not precise. What nodes? What packets?

Reviewer 2: Can you please address the observation above?

The following 2 diagrams show the packet flow for both egress and return traffic when using direct
routing mode. Note that the return packet is not going over a tunnel, while forwarding traffic is. 

.. image:: images/cilium_egress_gateway_flow.png
   :scale: 70%

.. image:: images/cilium_egress_gateway_return_flow.png
   :scale: 70%

> Create external service
> =======================

Reviewer 2: ```suggestion
Create an External Service
==========================
```

Create an nginx service on a Linux node that is external to the existing Kubernetes
cluster, and use it as the destination of the egress traffic.

.. code-block:: shell-session

    $ # Install and start nginx
    $ sudo apt install nginx
    $ sudo systemctl start nginx

>     $ # Make sure the service is started and listen on port :80

Reviewer 2: ```suggestion
    $ # Make sure the service is started and listens on port :80
```

    $ sudo systemctl status nginx
    ● nginx.service - A high performance web server and a reverse proxy server
    Loaded: loaded (/lib/systemd/system/nginx.service; enabled; vendor preset: enabled)
    Active: active (running) since Sun 2021-04-04 21:58:57 UTC; 1min 3s ago
    [...]
    $ curl http://192.168.33.13:80  # Assume 192.168.33.13 is the external IP of the node
    [...]
    <title>Welcome to nginx!</title>
    [...]

> Create client pods

Reviewer 2: ```suggestion
Create Client Pods
```

==================

Deploy a client pod from which egress traffic is generated, and try to access previous

> external service

Reviewer 2: ```suggestion
external service.
```

Kubernetes cluster.

.. code-block:: shell-session

    $ tail /var/log/nginx/access.log
    192.168.33.11 - - [04/Apr/2021:22:06:57 +0000] "GET / HTTP/1.1" 200 612 "-" "Wget/1.19.4 (linux-gnu)"

In the previous example, the busybox is running on the node ``192.168.33.11``, so the result makes sense.

Also apply a label to the default namespace which will be used in the future.

.. code-block:: shell-session

    $ kubectl label ns/default ns=default

Configure Egress IPs
====================

Deploy the following deployment to assign additional egress IP to the gateway node. The node that runs the
pod will have additional IP addresses configured on the external interface, and become the egress gateway.
In the following example, ``192.168.33.100`` and ``192.168.33.101`` becomes the egress IP which can be
consumed by Egress NAT Policy.

.. literalinclude:: ../../examples/kubernetes-egress-gateway/egress-ip-deployment.yaml

Create Egress NAT Policy
========================

Apply the following Egress NAT Policy, which basically means: when the pod is running in the namespace
with label ``ns=default`` and the pod itself has label ``run=client``, if it's trying to talk to IP CIDR

> ``192.168.33.13/32``, use egress IP ``192.168.33.100``. In this example, it tells cilium to forward the

Reviewer 2: ```suggestion
``192.168.33.13/32``, use egress IP ``192.168.33.100``. In this example, it tells Cilium to forward the
```

> packet from pod ``client`` to the gateway node with egress IP ``192.168.33.100``, and masqurade with that

Reviewer 2: ```suggestion
packet from pod ``client`` to the gateway node with egress IP ``192.168.33.100``, and masquerade with that
```

IP.

.. literalinclude:: ../../examples/kubernetes-egress-gateway/egress-nat-policy.yaml

Let's switch back to the client pod and verify it works.

.. code-block:: shell-session

    $ kubectl exec -it client --image=busybox
    If you don't see a command prompt, try pressing enter.
    / # wget http://192.168.33.13:80 -o /dev/null

Verify access log from nginx node that the request is coming from egress IP now instead of one of the
nodes in Kubernetes cluster.

>     .. code-block:: shell-session

Reviewer 2: Wrong indent level.

Author: This is fixed.

        $ tail /var/log/nginx/access.log
        [...]
        192.168.33.100 - - [04/Apr/2021:22:06:57 +0000] "GET / HTTP/1.1" 200 612 "-" "Wget/1.19.4 (linux-gnu)"

> 
> 

Reviewer 2: ```suggestion
```

Author: Fixed. Thanks!
